File: service_appointment/service_appointment/page/service_team_schedule/service_team_schedule.js
frappe.pages['service-team-schedule'].on_page_load = function (wrapper) {
	var me = this;
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Service Team Schedule',
		single_column: true
	});

	frappe.breadcrumbs.add("Alqallaf");


	this.date_range = page.add_field({
		fieldtype: 'DateRange',
		fieldname: 'selected_date_range',
		label: __("Date Range"),
		placeholder: "Date Range",
		default: [frappe.datetime.month_start(), frappe.datetime.month_end()],
		input_class: 'input-sm',
		reqd: 1
	});

	this.show_schedule = page.add_field({
		fieldname: "show_schedule",
		label: __("Show Schedule"),
		fieldtype: "Button",
		click: (e) => {
			var errors = [];

			if (!this.date_range.value) {
				errors.push('- Date Range');
			}

			if (errors.length > 0) {
				frappe.msgprint("Please enter:<br>" + errors.join("<br>"), "Error");
			} else {
				show_schedule(this.date_range.value);
			}

		}
	});

	this.update = page.add_field({
		fieldname: "update",
		label: __("Update"),
		fieldtype: "Button",
		click: (e) => {
			var changed_data = [];

			$('.checkboxes').each(function (i, e) {
				var date = $(e).data('date');
				var team = $(e).data('team');
				var hour = $(e).data('hour');
				var new_value = e.checked;

				if ($(e).data('value') != e.checked) {
					changed_data.push({
						date: $(e).data('date'),
						team: $(e).data('team'),
						hour: $(e).data('hour'),
						value: (e.checked ? 1 : 0)
					});
				}
			});

			frappe.call({
				"method": "service_appointment.service_appointment.page.service_team_schedule.service_team_schedule.update_schedule",
				args: {
					data: changed_data
				},
				freeze: true,
				freeze_message: __("Updating Schedule..."),
				callback: function (r) {
					if (r.message) {
						frappe.show_alert({ message: __(r.message), indicator: 'green' });
					}
					show_schedule(cur_page.page.date_range.value);
				}
			});
		}
	});

	this.show_schedule.$input.addClass('btn-info');
	this.update.$input.addClass('btn-primary');
};

var show_schedule = function (date_range) {
	frappe.call({
		"method": "service_appointment.service_appointment.page.service_team_schedule.service_team_schedule.get_schedule_records",
		args: {
			date_range: date_range
		},
		freeze: true,
		freeze_message: __("Getting Schedule..."),
		callback: function (r) {
			var data = r.message;
			var hours = [
				{ name: 'hour_00', hour: '00' },
				{ name: 'hour_01', hour: '01' },
				{ name: 'hour_02', hour: '02' },
				{ name: 'hour_03', hour: '03' },
				{ name: 'hour_04', hour: '04' },
				{ name: 'hour_05', hour: '05' },
				{ name: 'hour_06', hour: '06' },
				{ name: 'hour_07', hour: '07' },
				{ name: 'hour_08', hour: '08' },
				{ name: 'hour_09', hour: '09' },
				{ name: 'hour_10', hour: '10' },
				{ name: 'hour_11', hour: '11' },
				{ name: 'hour_12', hour: '12' },
				{ name: 'hour_13', hour: '13' },
				{ name: 'hour_14', hour: '14' },
				{ name: 'hour_15', hour: '15' },
				{ name: 'hour_16', hour: '16' },
				{ name: 'hour_17', hour: '17' },
				{ name: 'hour_18', hour: '18' },
				{ name: 'hour_19', hour: '19' },
				{ name: 'hour_20', hour: '20' },
				{ name: 'hour_21', hour: '21' },
				{ name: 'hour_22', hour: '22' },
				{ name: 'hour_23', hour: '23' }
			];

			var html = frappe.render_template('service_team_schedule', {
				data: data,
				hours: hours
			});

			$('.layout-footer').removeClass('hide');
			$('.layout-footer').html(html);
		}
	});
};

$(document).on('click', '.check-all', function () {
	if ($(this).prop("checked") == true) {
		$(this).parents('tr').find('.checkboxes').prop("checked", "checked");
	} else if ($(this).prop("checked") == false) {
		$(this).parents('tr').find('.checkboxes').prop("checked", "");
	}
});

File: service_appointment/service_appointment/page/service_team_schedule/service_team_schedule.py
from __future__ import unicode_literals, print_function
import frappe
import json
from pprint import pprint
from frappe.utils import flt

@frappe.whitelist()
def get_schedule_records(date_range):
    teams = frappe.get_all('Team', fields=['name', 'team_name'], order_by='team_name asc')
    date_range = json.loads(date_range)
    date_list = []
    out = []

    for count in range(frappe.utils.date_diff(date_range[1], date_range[0])+1):
        date_list.append(frappe.utils.add_days(date_range[0], count))

    emp_string = ''
    if teams:
        teams_string = "'" + "', '".join([team.name for team in teams]) + "'"
        
        records = frappe.db.sql("""SELECT sta.*, team.team_name, team.name as team_docname
            FROM `tabService Team Availability` sta
            LEFT JOIN `tabTeam` team ON sta.team = team.name
            WHERE sta.team in ({teams_string})
            AND sta.date >= %(start_date)s
            AND sta.date <= %(end_date)s
            ORDER BY team.team_name asc
        """.format(teams_string=teams_string), {'start_date': date_range[0], 'end_date': date_range[1]}, as_dict=1)

        for date in date_list:
            date_records = []

            for team in teams:
                exist = False
                for record in records:
                    if team.team_name == record.team_name and date == frappe.utils.get_date_str(record.date):
                        date_records.append(record)
                        exist = True
            
                if not exist:
                    date_records.append({
                        'team_docname': team.name,
                        'team_name': team.team_name,
                        'date': date,
                        'hour_00': 0,
                        'hour_01': 0,
                        'hour_02': 0,
                        'hour_03': 0,
                        'hour_04': 0,
                        'hour_05': 0,
                        'hour_06': 0,
                        'hour_07': 0,
                        'hour_08': 0,
                        'hour_09': 0,
                        'hour_10': 0,
                        'hour_11': 0,
                        'hour_12': 0,
                        'hour_13': 0,
                        'hour_14': 0,
                        'hour_15': 0,
                        'hour_16': 0,
                        'hour_17': 0,
                        'hour_18': 0,
                        'hour_19': 0,
                        'hour_20': 0,
                        'hour_21': 0,
                        'hour_22': 0,
                        'hour_23': 0
                    })
            
            out.append({
                'date_display': frappe.utils.get_datetime(date).strftime('%A, %-d %B %Y'),
                'records': date_records
            })

    return out

@frappe.whitelist()
def update_schedule(data):
    import json

    changed_data = json.loads(data)
    data_dict = frappe._dict()

    for item in changed_data:
        data_dict.setdefault((item['date'], item['team']), []).append({'name': item['hour'], 'value': item['value']})

    for date, team in data_dict:
        hours = data_dict.get((date, team), [])
        hours_string = [(str(hour['name']) + ' = ' + str(hour['value'])) for hour in hours]
        hours_string = ', '.join(hours_string)
        if frappe.db.exists({'doctype': 'Service Team Availability', 'date': date, 'team':team}):
            frappe.db.sql("""UPDATE `tabService Team Availability`
                SET {hours_string}
                WHERE date = %(date)s
                AND team = %(team)s""".format(hours_string=hours_string), {'date': date, 'team': team})
        else:
            new_doc = frappe.new_doc('Service Team Availability')
            new_doc.set('date', date)
            new_doc.set('team', team)
            for hour in hours:
                new_doc.set(hour['name'], hour['value'])
            new_doc.save()

    return "Schedule Updated Successfully"

@frappe.whitelist()
def schedule(date, employee, shift_type):
    new_doc = frappe.new_doc("Shift Assignment")
    new_doc.date = date
    new_doc.employee = employee
    new_doc.shift_type = shift_type
    new_doc.save()
    new_doc.submit()

    return "Shift Assigned Successfully"

@frappe.whitelist()
def cancel_shift_assignment(shift_assignment):
    old_doc = frappe.get_doc("Shift Assignment", shift_assignment)
    old_doc.cancel()

    return "Shift Cancelled Successfully"